include/hip/hcc_detail/hip_runtime_api.h

```diff
@@ -2368,6 +2468,103 @@ hipError_t hipModuleGetFunction(hipFunction_t* function, hipModule_t module, con
 
 hipError_t hipFuncGetAttributes(hipFuncAttributes* attr, const void* func);
 
+struct Agent_global {
+    std::string name;
+    hipDeviceptr_t address;
+    uint32_t byte_cnt;
+};
+#if defined(__cplusplus)
+} // extern "C"
+#endif
+
+namespace hip_impl {
+hsa_executable_t executable_for(hipModule_t);
+const std::string& hash_for(hipModule_t);
+
+template<typename ForwardIterator>
+std::pair<hipDeviceptr_t, std::size_t> read_global_description(
+    ForwardIterator f, ForwardIterator l, const char* name) {
+    const auto it = std::find_if(f, l, [=](const Agent_global& x) {
+        return x.name == name;
+    });
+
+    return it == l ?
+        std::make_pair(nullptr, 0u) : std::make_pair(it->address, it->byte_cnt);
+}
+
+std::vector<Agent_global> read_agent_globals(hsa_agent_t agent,
+                                             hsa_executable_t executable);
+hsa_agent_t this_agent();
+
+inline
+__attribute__((visibility("hidden")))
+hipError_t read_agent_global_from_module(hipDeviceptr_t* dptr, size_t* bytes,
+                                         hipModule_t hmod, const char* name) {
+    // the key of the map would the hash of code object associated with the
+    // hipModule_t instance
+    static std::unordered_map<
+        std::string, std::vector<Agent_global>> agent_globals;
```

Comment:
Could we only expose a static object that manages the lifetime of these globals in this header and move/hide all the implementation details in a .cpp file?  One motivation is shielding the algorithm from hazardous C++ environment of some users because this code would get generated during HIP's build rather than in the user environment.  As a bonus it would reduce compilation time of user applications and produce smaller code size.

Reply:
That's a brilliant idea and you should absolutely pursue it as a further development. Lots of bonuses, as you've observed!

Reply:
I'm not suggesting we do this in this PR but is there any architecture assumption made that would prevent moving all the machinery into a .cpp file impossible?